File: level1/21/mergeTwoSortedLists.py
class ListNode(object):
    def __init__(self, val=0, next=None, head =None):
        self.val = val
        self.next = next
        self.head = None
    def insert(self, val):
          newNode = ListNode(val)
          if(self.head):
                current = self.head
                while(current.next):
                  current = current.next
                current.next = newNode
          else:
                self.head = newNode
    def printList(self):
        current = self.head
        while(current):
            print(current.val, end=' ')
            current = current.next

class Solution(object):
    def mergeTwoLists(self, list1, list2):   
        head_ptr = temp_ptr = ListNode()
        while list1 or list2:
            if list1 and ( not list2 or list1.val <= list2.val):
                temp_ptr.next = ListNode(list1.val)
                list1 = list1.next
            else:
                temp_ptr.next = ListNode(list2.val)
                list2 = list2.next
            temp_ptr = temp_ptr.next    
        return head_ptr.next

File: level1/21/mergeTwoSortedLists.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
	int val;
 	struct ListNode *next;
};

struct ListNode*	mergeLists(struct ListNode* list1, struct ListNode* list2) {
	struct ListNode* temp = list1;
	while (temp->next != NULL)
		temp = temp->next;
	temp->next = list2;
	return (list1);
}

void	swap(int *a, int *b) {
	int c;
	c = *a;
	*a = *b;
	*b = c;
}

struct ListNode*	sortList(struct ListNode* head) {
	struct ListNode* temp;

	if (!head)
		return (NULL);
	temp = head;
	while(head->next != NULL)
	{
		if (head->val > head->next->val)
		{
			while (head->val > head->next->val){
				swap(&head->val, &head->next->val);
			}
			head = temp;
		}
		else
			head = head->next;
	}
	return (temp);
}

struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {
	if (!list1)
		return (sortList(list2));
	else if (!list2)
		return (sortList(list1));
	else if (!list1 && !list2)
		return (NULL);
	struct ListNode* head = mergeLists(list1, list2);
	return (sortList(head));
}

int	main(void)
{
	struct ListNode *sorted;
	struct ListNode c = { 4, NULL};
	struct ListNode b = { 2, &c };
	struct ListNode a = { 1, &b };
 	struct ListNode *list1 = &a;
	struct ListNode l3 = { 4, NULL};
	struct ListNode l2 = { 3, &l3 };
	struct ListNode l1 = { 1, &l2 };
 	struct ListNode *list2 = &l1;
	sorted = mergeTwoLists(list1, list2);
	while (sorted->next != NULL)
	{
		printf("%d ", sorted->val);
		sorted = sorted->next;
	}
	return (0);
}